crypto-mcl.ts

```diff
@@ -0,0 +1,188 @@
+
+        const R1 = mcl.mul(mcl.baseG2(), r1);
+        const H1inc = this.infoNonceCounter(info, proofEntry, counter);
+        const Z1 = mcl.pairing(mcl.mul(H1inc, r1), entry);
+        const R2 = mcl.mul(mcl.baseG2(), r2);
+        const Z2 = mcl.pairing(mcl.mul(H1inc, r2), entry);
+        const K = mcl.xor(k, mcl.hashT(Z2));
+
+        return {R1, Z1, R2, K, c, nonce};
+    }
+
+    /**
+     * Convenience method to calculate the hash-to-G1 of the info, nonce, and counter.
+     * @param info
+     * @param nonce
+     * @param counter
+     */
+    static infoNonceCounter(info: string, nonce: Uint8Array, counter: number): mcl.G1 {
+        const counterStr = counter.toString();
+        const infoNonceCnt = new Uint8Array(info.length + nonce.length + counterStr.length);
+        infoNonceCnt.set(from_string(info));
+        infoNonceCnt.set(nonce, info.length);
+        infoNonceCnt.set(from_string(counterStr), info.length + nonce.length);
+        return mcl.hashAndMapToG1(infoNonceCnt);
+    }
+
+    /**
+     * Generates a pre-trace by the location owner, one for every slot of possible infection.
+     * @param mtr master trace record from the location
+     * @param cnt hours since the unix epoch
+     * @return an IPreTrace for the health authority
+     */
+    static genPreTrace(mtr: IMasterTrace, cnt: number): IPreTrace{
+        const {info, nonce, mskv, mskHAEnc: ctxt} = mtr;
+        const Tprime = mcl.mul(this.infoNonceCounter(info, nonce, cnt), mskv);
+        return {Tprime, info, nonce, ctxt};
+    }
+
+    /**
+     * Generates an anonymous trace that can only be used by people having scanned the QRentry code
+     * @param skH the health authorities private key
+     * @param cnt hours since the unix epoch
+     * @param ptr from genPreTrace
+     * @param proofPTR is ignored for the moment
+     * @return tracing information to be stored
+     */
+    static genTrace(skH: Uint8Array, cnt: number, ptr: IPreTrace, proofPTR?: Uint8Array): mcl.G1{
+        const {Tprime, info, nonce, ctxt} = ptr;
+        const mskHA = new mcl.Fr();
+        const pkH = crypto_scalarmult_base(skH)
+        mskHA.deserialize(crypto_box_seal_open(ctxt, pkH, skH));
+        const B = this.infoNonceCounter(info, nonce, cnt);
+        return mcl.add(Tprime, mcl.mul(B, mskHA));
+    }
+
+    /**
+     * Tries to match a user record against a trace.
+     * @param rec one of the user records created by scan
+     * @param tr one of the traces received by the health authority
+     * @return the aux string (if empty will be "") or undefined if no match occurred
+     */
+    static match(rec: IUserRecord, tr: mcl.G1): (string|undefined){
+        const Z1prime = mcl.pairing(tr, rec.R1);
+        if (!Z1prime.isEqual(rec.Z1)){
+            return undefined;
+        }
+        try {
+            const k = mcl.xor(rec.K, mcl.hashT(mcl.pairing(tr, rec.R2)));
+            const aux = crypto_secretbox_open_easy(rec.c, rec.nonce, k);
```

Comment:
here as well: please check return value. If decryption fails, better explicitly return undefined (instead of relying on `to_string` to fail in a proper manner).

Reply:
in fact here I catch it: if the `cs_open_easy` fails, it throws an error, which is caught, and then the method returns `undefined`. I did not do it with the other methods, as the whitepaper did not mention that it can return `_I_`...